cmake/Modules/PkgBuildConfig.cmake

```diff
@@ -0,0 +1,101 @@
+# pkg-config file generation
+#
+# Uses the following globals:
+# - PKG_BUILD_PREFIX: the build location (aka prefix). Defaults to CMAKE_INSTALL_PREFIX
+# - PKG_BUILD_LIBDIR: the libdir location. Defaults to ${prefix}/lib.
+# - PKG_BUILD_INCLUDEDIR: the includedir location. Defaults to ${prefix}/include.
+#
+function(pkg_build_config)
+    set(options)
+    set(oneValueArgs NAME DESCRIPTION VERSION FILENAME LIBS_SELF)
+    set(multiValueArgs LIBS PRIVATE_LIBS REQUIRES CFLAGS)
+
+    cmake_parse_arguments(PKGCONFIG "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if (NOT DEFINED PKGCONFIG_FILENAME AND DEFINED PKGCONFIG_NAME)
+        set(PKGCONFIG_FILENAME ${PKGCONFIG_NAME})
+    endif()
+    if (NOT DEFINED PKGCONFIG_FILENAME)
+        message(FATAL_ERROR "Missing FILENAME argument")
+    endif()
+    set(PKGCONFIG_FILE "${PROJECT_BINARY_DIR}/${PKGCONFIG_FILENAME}.pc")
+
+    if (NOT DEFINED PKGCONFIG_DESCRIPTION)
+        message(FATAL_ERROR "Missing DESCRIPTION argument")
+    endif()
+
+    if (NOT DEFINED PKGCONFIG_VERSION)
+        message(FATAL_ERROR "Missing VERSION argument")
+    endif()
+
+    if (DEFINED PKG_BUILD_PREFIX)
+        set(PKGCONFIG_PREFIX ${PKG_BUILD_PREFIX})
+    else()
+        set(PKGCONFIG_PREFIX ${CMAKE_INSTALL_PREFIX})
+    endif()
+
+    if(DEFINED PKG_BUILD_LIBDIR)
+        if (IS_ABSOLUTE ${PKG_BUILD_LIBDIR})
+            set(PKGCONFIG_LIBDIR ${PKG_BUILD_LIBDIR})
+        else()
+            set(PKGCONFIG_LIBDIR "\${prefix}/${PKG_BUILD_LIBDIR}")
+        endif()
+    else()
+        set(PKGCONFIG_LIBDIR "\${prefix}/lib")
+    endif()
+
+    if(DEFINED PKG_BUILD_INCLUDEDIR)
+        if (IS_ABSOLUTE ${PKG_BUILD_INCLUDEDIR})
+            set(PKGCONFIG_INCLUDEDIR ${PKG_BUILD_INCLUDEDIR})
+        else()
+            set(PKGCONFIG_INCLUDEDIR "\${prefix}/${PKG_BUILD_INCLUDEDIR}")
+        endif()
+    else()
+        set(PKGCONFIG_INCLUDEDIR "\${prefix}/include")
+    endif()
+
+    file(WRITE "${PKGCONFIG_FILE}"
+        "prefix=${PKGCONFIG_PREFIX}\n"
+        "libdir=${PKGCONFIG_LIBDIR}\n"
+        "includedir=${PKGCONFIG_INCLUDEDIR}\n"
+        "\n"
+        "Name: ${PKGCONFIG_NAME}\n"
+        "Description: ${PKGCONFIG_DESCRIPTION}\n"
+        "Version: ${PKGCONFIG_VERSION}\n"
+    )
+
+    if(NOT DEFINED PKGCONFIG_LIBS_SELF)
+        set(PKGCONFIG_LIBS_SELF "${PKGCONFIG_FILE}")
+    endif()
+
+    if(NOT DEFINED PKGCONFIG_LIBS)
+        set(PKGCONFIG_LIBS "-l${PKGCONFIG_LIBS_SELF}")
```

Comment:
This seems a bit weird to me. So you're setting the PKGCONFIG_LIBS variable to the filename of the pkgconfig file?

Reply:
It's part of a multi-step fallback (which is obviously undocumented :wink:). The goal is to have sane defaults (ie. the project, dylib & pkg-config identifier are all equal), but allow customization. So `PKGCONFIG_LIBS` is the underlying, substitued-in value, but we need to preserve what's in it (eg. mbedtls has a dependency on pthreads, which isn't covered by pkg-config), we have somewhat non-matching names — we're known as `libgit2` to pkg-config, but our `LIBS_SELF` name for linking is `-lgit2`.

In short, it's the reverse, the filename defaults to the pkg-config `NAME`, unless `FILENAME` is passed. The `-l$self` flag defaults to `SELF_LIBS`, `FILENAME` or `NAME`, in that order.

I *could* have considered that `NAME` must be a known target (which would allow to `GET_TARGET_PROPERTY`), but I didn't. It's just a dump "collect, customize, and write" wrapper.

Reply:
Fair enough, thanks for explaining! :)